### PR TITLE
Refactor to remove code duplication and cleanup data streaming custom cmds.

### PR DESCRIPTION
**Pull request by edgarsilva**

### CODE REVIEW

File: lib/devices/custom.js [hunk 1]
```diff
@@ -155,30 +155,23 @@ module.exports = function custom(device) {
...
       m: 1,
-      mask1: mask1,
+      mask1: this._mergeMasks("mask1", args.mask1, args.remove),
       pcnt: 0,
-      mask2: mask2
+      mask2: this._mergeMasks("mask2", args.mask2, args.remove)
+    };
+
+    var callback = function(data) {
+      console.log("CHECKPOINT ==>", args.event);
```
Comment (stewart) [inline lib/devices/custom.js]: Should this be here?

Comment (edgarsilva) [inline lib/devices/custom.js]: There was a commit missing, just pushed.

**Comment by stewart** [comment]
Per the `_`-prefixed commands, I'd prefer if they could be extracted out of the mutator function (like `hexToRgb`) - we don't want to pollute the `device` any more than necessary, and if it's intended to be a private method it should be marked as such (`@private`) such that the JSDoc importer doesn't grab it.

Other than that this looks pretty good.

File: lib/devices/custom.js [hunk 2]
```diff
@@ -155,30 +155,29 @@ module.exports = function custom(device) {
...
+      mask2: this._mergeMasks("mask2", args.mask2, args.remove)
+    };
+
+    var fields = args.fields;
+
+    var callback = function(data) {
+      var params = {};
+      for (var i = 0; i < fields.length; i++) {
+        params[fields[i]] = data[fields[i]];
+      }
```
Comment (stewart) [inline lib/devices/custom.js]: Can we just use `fields.forEach` here?

**Comment by edgarsilva** [comment]
Sure, which prefixed commands are you talking about? _mergeMasks and _streamData? suggestions of where to put them? I guess stream data does not need to be a `_` preffix commands since it is the meat and bones. What do you think?

File: lib/devices/custom.js (hunk 2 above)
Comment (edgarsilva) [inline lib/devices/custom.js]: do not want to add yet another callback.

**Comment by stewart** [comment]
Here's my proposed diff -> https://gist.github.com/stewart/e0c48cb2f14cf410c278

Let me know if I've misread something, or gotten something wrong.

If you like it, I can commit and push it.

**Comment by deadprogram** [comment]
:+1:

**Comment by edgarsilva** [comment]
Most of it looks great, still not sold in using `forEach` instead of `for loop` and adding yet another callback and context level to the execution just for that. Looks to me like `for` loop is best here to avoid it, at least for this in this context for the params setup, also quicker.

Other than that I think is solid.

**Comment by stewart** [comment]
Updated.

**Comment by edgarsilva** [comment]
Who's merging?

**Comment by stewart** [comment]
standoff

**Pull request merged**